### PR TITLE
 Cleaning nats subscriptions on session destroy

### LINKED ISSUE #533: session-create topic deduplication
Following code introduces topic duplication:
```
func (receiver *receiverNATS) Respond(consumer communication.RequestConsumer) error {
...	
	_, err := receiver.connection.Subscribe(requestTopic, messageHandler)
```

Since in case of consumer timeout, session is not cleaned up and topic is not destroyed.
This leads to memory leaks and potential security threat (since topic can be reused). Investigate other cases when there are topic leaks.

Each new session-create requests introduces new consumer for the same topic: 

`
2018-11-12T13:29:46.238181 [Debug] [NATS.Receiver] Request '0x207bab3168645844c8f53a5bc220557e622701fb.dialog-create' response: {"payload":{"reason":200,"reasonMessage":"OK"},"signature":"F/JEOkqoyDQpraaNouFiDV6nCoMFcTVsICLQ9FFUyA4sfwmodk2h1oNpw02ssU7tm1nKIK9CCCV0KPNHQO9CpQE="}
2018-11-12T13:29:46.239923 [Debug] [NATS.Receiver] Request '0x207bab3168645844c8f53a5bc220557e622701fb.0x8e586ca51c0f0e83564747bfd4d73bcd90119fc7.session-create' received: {"payload":{"proposal_id":1},"signature":"nThhz+bAn91+sNi7gd1G0vILRLzRs5fUxAlYmvRaJYwDnLfqrsjE+LtZnYHqZlVVpBSLP7Zt8Vue7ZAYWVax8wE="}
2018-11-12T13:29:46.240062 [Debug] [NATS.Receiver] Request '0x207bab3168645844c8f53a5bc220557e622701fb.0x8e586ca51c0f0e83564747bfd4d73bcd90119fc7.session-create' received: {"payload":{"proposal_id":1},"signature":"nThhz+bAn91+sNi7gd1G0vILRLzRs5fUxAlYmvRaJYwDnLfqrsjE+LtZnYHqZlVVpBSLP7Zt8Vue7ZAYWVax8wE="}`

### PR DESCRIPTION
**Pull request by soffokl**
Closes #533

### CODE REVIEW

File: communication/nats/receiver.go [hunk 1]
```diff
@@ -66,12 +70,28 @@ func (receiver *receiverNATS) Receive(consumer communication.MessageConsumer) er
 		}
 	}
 
-	_, err := receiver.connection.Subscribe(messageTopic, messageHandler)
+	receiver.mu.Lock()
+	defer receiver.mu.Unlock()
+
+	subs, err := receiver.connection.Subscribe(messageTopic, messageHandler)
```
Comment (vkuznecovas) [inline communication/nats/receiver.go]: might be a nitpick, but it returns a sub, not subs(the s on the end indicates a plural, in my head at least)

File: communication/nats/receiver.go [hunk 2]
```diff
@@ -111,17 +131,22 @@ func (receiver *receiverNATS) Respond(consumer communication.RequestConsumer) er
 		}
 	}
 
+	receiver.mu.Lock()
+	defer receiver.mu.Unlock()
+
+	if subs, ok := receiver.subs[requestTopic]; ok && subs.IsValid() {
```
Comment (vkuznecovas) [inline communication/nats/receiver.go]: might be a nitpick, but it returns a sub, not subs(the s on the end indicates a plural, in my head at least)

File: communication/nats/receiver.go (hunk 1 above)
Comment (soffokl) [reply communication/nats/receiver.go]: correct, fixed it.

File: communication/nats/receiver.go (hunk 2 above)
Comment (soffokl) [reply communication/nats/receiver.go]: correct, fixed it.

File: communication/nats/receiver.go [hunk 3]
```diff
@@ -66,12 +70,28 @@ func (receiver *receiverNATS) Receive(consumer communication.MessageConsumer) er
...
+	receiver.subs[messageTopic] = subscription
+	return nil
+}
+
+func (receiver *receiverNATS) Unsubscribe() error {
+	receiver.mu.Lock()
+	defer receiver.mu.Unlock()
 
+	for topic, s := range receiver.subs {
+		if err := s.Unsubscribe(); err != nil {
```
Comment (tadovas) [inline communication/nats/receiver.go]: If one of subscribers returns error, do we really want to skip unsubscribing of others? Maybe just log error and continue? Or maybe unsubscribe doesn't need to return error at all?

Comment (soffokl) [reply communication/nats/receiver.go]: Changed it to just log errors.

**Pull request merged**